lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -1378,50 +1380,66 @@ private static function sort_start_ascending( $a, $b ) {
+	 * @return string|boolean|null Value of enqueued update if present, otherwise false.
 	 */
-	private function extract_attribute_value_from_lexical_update( $name ) {
-		$comparable = strtolower( trim( $name ) );
-
-		if ( ! isset( $this->lexical_updates[ $comparable ] ) ) {
-			return null;
+	private function get_enqueued_attribute_value( $comparable_name ) {
+		if ( ! isset( $this->lexical_updates[ $comparable_name ] ) ) {
+			return false;
 		}
 
-		$attribute = trim( $this->lexical_updates[ $comparable ]->text );
-		// Has the attribute been removed?
-		if ( '' === $attribute ) {
+		$enqueued_text = $this->lexical_updates[ $comparable_name ]->text;
+
+		// Removed attributes erase the entire span.
+		if ( '' === $enqueued_text ) {
 			return null;
 		}
 
-		/**
-		 * When the lexical update contains just the attribute name,
-		 * the value becomes a boolean true. For example:
-		 *
-		 * ```php
-		 * $p = new WP_HTML_Tag_Processor('<input type="checkbox" />');
-		 * $p->set_attribute('checked', true);
-		 * // The lexical update contains just the string "checked"
+		/*
+		 * Boolean attribute updates are just the attribute name without a corresponding value.
 		 *
-		 * $p->get_attribute('checked');
-		 * // returns true thanks to the condition below
+		 * This value might differ from the given comparable name in that there could be leading
+		 * or trailing whitespace, and that the casing follows the name given in `set_attribute`.
 		 *
-		 * echo $p->get_updated_html();
-		 * // <input type="checkbox" checked />
+		 * Example:
+		 * ```
+		 *     $p->set_attribute( 'data-TEST-id', 'update' );
+		 *     'update' === $p->get_enqueued_attribute_value( 'data-test-id' );
 		 * ```
+		 *
+		 * Here we detect this based on the absence of the `=`, which _must_ exist in any
+		 * attribute containing a value, e.g. `<input type="text" enabled />`.
+		 *                                            ¹           ²
+		 *                                       1. Attribute with a string value.
+		 *                                       2. Boolean attribute whose value is `true`.
 		 */
-		if ( $attribute === $comparable ) {
+		$equals_at = strpos( $enqueued_text, '=' );
+		if ( false === $equals_at ) {
 			return true;
 		}
 
-		// $attribute is a 'name="value"' string, so we extract the value.
-		$value = substr( $attribute, strlen( $comparable ) + 2, -1 );
-		return html_entity_decode( $value );
+		/*
+		 * Finally, a normal update's value will appear after the `=` and
+		 * be double-quoted, as performed incidentally by `set_attribute`.
+		 *
+		 * e.g. `type="text"`
+		 *           ¹²    ³
+		 *        1. Equals is here (E).
+		 *        2. Double-quoting starts one after the equals sign (E + 1).
+		 *        3. Double-quoting ends at the last character in the update (Total Length - E - 3 syntax characters).
+		 */
+		$enqueued_value = substr( $enqueued_text, $equals_at + 2, strlen( $enqueued_text ) - $equals_at - 3 );
```

Comment:
Can't we use

```suggestion
		 *        3. Double-quoting ends at the last character in the update.
		 */
		$enqueued_value = substr( $enqueued_text, $equals_at + 2, -1 );
```

to point to the end of the string, minus one character? I feel like that's easier to follow than the `strlen( $enqueued_text ) - $equals_at - 3` computation 😬 

`substr` [docs](https://www.php.net/manual/en/function.substr.php) for the `$length` arg:

> If `length` is given and is negative, then that many characters will be omitted from the end of string

Reply:
very good idea! why didn't I think of that 😄